### PR TITLE
Make linuxkit Hook services resilient so that the getty always comes up:

### PR DESCRIPTION
**Pull request by jacobweinstock**
## Description

Hook-bootkit requires a network connection to pull down the tink-worker image. The previous solution was to make sure networking was up and running via a DHCP client in the linuxkit "onboot" section. This meant that any issues with the DHCP onboot container would cause the getty, and all services, to never start. This made troubleshooting nearly impossible.

This PR makes Hook-bootkit (and hook-docker) retry all failures. This allows us to remove the DHCP "onboot" container. This should allow the getty service to always start.

## Why is this needed

Fixes: #

## Checklist:

I have:

- [ ] updated the documentation and/or roadmap (if required)
- [ ] added unit or e2e tests
- [ ] provided instructions on how to upgrade

### CODE REVIEW

**Review by chrisdoherty4 — APPROVED** [review]

File: hook-bootkit/main.go [hunk 1]
```diff
@@ -71,11 +116,74 @@ func main() {
...
+		return err
+	}
+
+	if err = out.Close(); err != nil {
+		log.Error(err, "error when closing image pull logs")
 	}
 
+	cs, err := cli.ContainerList(ctx, types.ContainerListOptions{All: true})
+	if err != nil {
+		return fmt.Errorf("error listing containers used for checking for an existing tink-worker container: %w", err)
```
Comment (chrisdoherty4) [inline hook-bootkit/main.go:174]:
```suggestion
		return fmt.Errorf("isting containers used for checking for an existing tink-worker container: %w", err)
```

File: hook-bootkit/main.go [hunk 2]
```diff
@@ -71,11 +116,74 @@ func main() {
...
+	}
+	for _, c := range cs {
+		for _, n := range c.Names {
+			if n == "/tink-worker" {
+				log.Info("Removing existing tink-worker container")
+				if err := cli.ContainerRemove(ctx, c.ID, types.ContainerRemoveOptions{Force: true}); err != nil {
+					return fmt.Errorf("error removing existing tink-worker container: %w", err)
```
Comment (chrisdoherty4) [inline hook-bootkit/main.go:181]:
```suggestion
					return fmt.Errorf("removing existing tink-worker container: %w", err)
```

File: hook-bootkit/main.go [hunk 3]
```diff
@@ -110,78 +218,36 @@ func main() {
...
+		return fmt.Errorf("error creating tink-worker container: %w", err)
 	}
 
-	authStr := base64.URLEncoding.EncodeToString(encodedJSON)
-
-	pullOpts := types.ImagePullOptions{
-		RegistryAuth: authStr,
+	log.Info("Starting tink-worker container")
+	if err := cli.ContainerStart(ctx, resp.ID, types.ContainerStartOptions{}); err != nil {
+		return fmt.Errorf("error starting tink-worker container: %w", err)
```
Comment (chrisdoherty4) [inline hook-bootkit/main.go:230]:
```suggestion
		return fmt.Errorf("starting tink-worker container: %w", err)
```

File: hook-bootkit/main.go [hunk 4]
```diff
@@ -110,78 +218,36 @@ func main() {
...
-	}
-	if err = backoff.Retry(imagePullOperation, backoff.WithMaxRetries(backoff.NewExponentialBackOff(), maxRetryAttempts)); err != nil {
-		panic(err)
-	}
-
-	if _, err = io.Copy(os.Stdout, out); err != nil {
-		panic(err)
+	// if tink-worker is not running return error so we try again
+	if err := checkContainerRunning(ctx, cli, resp.ID); err != nil {
+		return fmt.Errorf("error checking if tink-worker container is running: %w", err)
```
Comment (chrisdoherty4) [inline hook-bootkit/main.go:236]:
```suggestion
		return fmt.Errorf("checking if tink-worker container is running: %w", err)
```

File: hook-bootkit/main.go [hunk 5]
```diff
@@ -71,11 +116,74 @@ func main() {
...
+	for _, c := range cs {
+		for _, n := range c.Names {
+			if n == "/tink-worker" {
+				log.Info("Removing existing tink-worker container")
+				if err := cli.ContainerRemove(ctx, c.ID, types.ContainerRemoveOptions{Force: true}); err != nil {
+					return fmt.Errorf("error removing existing tink-worker container: %w", err)
+				}
+			}
+		}
+	}
```
Comment (chrisdoherty4) [inline hook-bootkit/main.go:185]: I think it'd be useful to either comment above this code block to the effect of `// Remove any existing Tink worker containers from previous runs` or put it in it own function.

**Review by chrisdoherty4 — APPROVED** [review]

File: hook-bootkit/main.go [hunk 6]
```diff
@@ -46,22 +52,61 @@ type tinkConfig struct {
 	noProxy       string
 }
 
-const maxRetryAttempts = 20
+// defaultLogger is a zerolog logr implementation.
+func defaultLogger(level string) logr.Logger {
+	zerolog.ErrorStackMarshaler = pkgerrors.MarshalStack
```
[the remark this answers is not in the document]
Comment (jacobweinstock) [reply hook-bootkit/main.go]: sure, i've removed this.

File: hook-bootkit/main.go (hunk 5 above)
Comment (jacobweinstock) [reply hook-bootkit/main.go]: sure thing, will update. thanks

**Comment by jacobweinstock** [comment]
Hey @chrisdoherty4 , thanks for the review. JFYI, this is just a refactor step. There is still much more work to be done. I will follow in the future with more refactoring.

**Review by chrisdoherty4 — APPROVED** [review]

**Pull request merged**